src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -355,34 +361,34 @@ private void TrackNullableStateForAssignment(BoundNode node, int targetSlot, Typ
                     return;
                 }
 
-                if (targetType.IsReferenceType)
+                bool isByRefTarget = IsByRefTarget(targetSlot);
+                if (targetSlot > 0)
                 {
-                    bool isByRefTarget = IsByRefTarget(targetSlot);
+                    if (targetSlot >= this.State.Capacity) Normalize(ref this.State);
+
+                    this.State[targetSlot] = isByRefTarget ?
+                        // Since reference can point to the heap, we cannot assume the value is not null after this assignment,
+                        // regardless of what value is being assigned. 
+                        (targetType.IsNullable == true) ? (bool?)false : null :
+                        !valueType?.IsNullable;
```

Comment:
Not related to this PR: Sorry if silly question, but I'm having trouble understanding the negative in `!valueType?.IsNullable`. 
From my understanding `valueType` is the null-state or result from the `value` expression. I would have thought that would directly flow to the target.

Reply:
Currently `LocalState` stores the value as not nullable rather than nullable. We should change the representation in `LocalState` or at least change the indexer on that type to hide the difference. Updated the comment on `LocalState`.